A29-xds-tls-security.md

```diff
@@ -0,0 +1,588 @@
+      .addService(new HostnameGreeter(hostname))
+      .build()
+      .start();
+```
+
+#### Go
+
+Use of XdsChannelCredentials:
+
+```Go
+   import (
+           xdscreds "google.golang.org/grpc/credentials/xds"
+   )
+
+   if creds, err := xdscreds.NewClientCredentials(xdscreds.ClientOptions{FallbackCreds: insecure.NewCredentials()}); err != nil {
+           log.Fatalf("Failed to create xDS client credentials: %v", err)
+   }
+   conn, err := grpc.Dial(*target, grpc.WithTransportCredentials(creds))
+```
+
+Use of XdsServerCredentials:
+
+```Go
+   import (
+           xdscreds "google.golang.org/grpc/credentials/xds"
+           "google.golang.org/grpc/xds"
+   )
+
+   if creds, err := xdscreds.NewServerCredentials(xdscreds.ServerOptions{FallbackCreds: insecure.NewCredentials()}); err != nil {
+           log.Fatalf("Failed to create xDS server credentials: %v", err)
+   }
+   server := xds.NewGRPCServer(grpc.Creds(creds))
+```
+
+### xDS Protocol
+
+xDS v3 is a pre-requisite for this proposal because the required fields were added to xDS v3. 
+gRPC uses CDS to configure client side security and LDS for server side security as described in the
+sections below. The sections describe how the CDS or LDS updates are processed by gRPC. Note that gRPC
+always validates all security configurations in CDS or LDS regardless of whether an XdsChannelCredentials
+or XdsServerCredentials is in force. If validation fails on an update, the update is NACKed.
+
+An unsupported field is normally ignored. However if ignoring a field compromises security, or
+if the unsupported field affects how we interpret other fields, we NACK the update when the
+field is present.
+
+When an update is accepted, the security configuration contained in that update is used if the
+application used Xds*Credentials.
+
+#### CDS for Client Side Security
+
+The CDS policy is the top level LB policy applied to all the connections under that policy.
+[A27:CDS][] describes the flow. The field [`transport_socket`][CL-TS] is used to extract the
+[`UpstreamTlsContext`][UTC] as described [here][CL-TS-comment].
+Note that we don't (currently) support [`transport_socket_matches`][CL-TS-matches].
+The security configuration extracted from the `UpstreamTlsContext` thus obtained
+is passed down to all the child policies and connections similar to how gRPC passes
+load balancer policy information down to child policies.
+
+[`common_tls_context`][CTC] in the `UpstreamTlsContext` contains the required security
+configuration. See below for [`CommonTlsContext`][CTC-type] processing.
+
+A secure client at a minimum requires a CA root certificate to be able to validate the server
+certificate in the TLS mode and this is provided by [ca_certificate_provider_instance][CCPI]
+inside the [`default_validation_context`][default_validation_context] of
+[combined_validation_context][CVC]. For backward compatibility gRPC will use the deprecated
```

Comment:
With the field cleanup, I had expected we'd use `validation_context`, not `combined_validation_context`. This seems to suggest using `combined_validation_context` with only `default_validation_context` set which defeats the point of `combined_validation_context`. I don't see a reason to do that from a compatibility standpoint.

Reply:
Yes I considered that. An envoy-compatible config from a control plane currently looks like this:

```
        "commonTlsContext": {
...
          "combinedValidationContext": {
            "defaultValidationContext": {
              "matchSubjectAltNames": [{
                "exact": "spiffe://foo.example.com/ns/example-grpc-server/sa/example-grpc-server"
              }]
            },
            "validationContextSdsSecretConfig": {
              "name": "validation_context_sds",
              "sdsConfig": {
                "path": "/etc/envoy/validation_context_sds_secret.yaml"
              }
            },
            "validationContextCertificateProviderInstance": {
              "instanceName": "google_cloud_private_spiffe",
              "certificateName": "ROOTCA"
            }
          },
...
        }
```

With the restructured fields and as described here it will change to:

```
        "commonTlsContext": {
...
          "combinedValidationContext": {
            "defaultValidationContext": {
              "matchSubjectAltNames": [{
                "exact": "spiffe://foo.example.com/ns/example-grpc-server/sa/example-grpc-server"
              }],
              "caCertificateProviderInstance": {
                 "instanceName": "foobar",
                 "certificateName": "ROOTCA"
              }
            },
            "validationContextSdsSecretConfig": {
              "name": "validation_context_sds",
              "sdsConfig": {
                "path": "/etc/envoy/validation_context_sds_secret.yaml"
              }
            }
          },
...
        }
```

Whereas what you are proposing wouldn't be possible if you still want to maintain Envoy compatibility. Because of the `oneof` the control plane can send only either `combined_validation_context` (which works for Envoy and gRPC) or `validation_context` (which wouldn't work for Envoy).

I guess I can make a small change here to look for `validation_context` (if present) otherwise `combined_validation_context.default_validation_context` to get the validation context information.

Reply:
Yeah, I think the `oneof` requires us to support `combined_validation_context` for Envoy compatibility, unfortunately.  But I think we can basically ignore the SDS part of it, because we have an existing rule that we NACK if `ca_certificate_provider_instance` is not set.

See my other comment with suggested replacement wording for this paragraph.